docker/src/main/java/clocker/docker/location/strategy/basic/GroupPlacementStrategy.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * Can be configured to require exclusive use of the host with the
+ * {@link #REQUIRE_EXCLUSIVE exclusive} ({@code docker.constraint.exclusive})
+ * option set to {@code true}; normally {@code false}.
+ *
+ * @since 1.1.0
+ */
+@Beta
+@ThreadSafe
+public class GroupPlacementStrategy extends AbstractDockerPlacementStrategy implements WithMutexes {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GroupPlacementStrategy.class);
+    private static final Object LOCK = new Object[0];
+
+    @SetFromFlag("exclusive")
+    public static final ConfigKey<Boolean> REQUIRE_EXCLUSIVE = ConfigKeys.newBooleanConfigKey(
+            "docker.constraint.exclusive",
+            "Whether the Docker host must be exclusive to this application; by default other applications can co-exist",
+            Boolean.FALSE);
+
+    @SetFromFlag("timeout")
+    public static final ConfigKey<Duration> STRATEGY_TIMEOUT = ConfigKeys.newDurationConfigKey(
+            "docker.strategy.timeout",
+            "How long to wait for other entities using the strategy",
+            Duration.minutes(20));
+
+    public static final ConfigKey<Map<String, Monitor>> MONITOR_MAP = ConfigKeys.newConfigKey(
+            new TypeToken<Map<String, Monitor>>() { },
+            "groupPlacementStrategy.map.monitors",
+            "A mapping from application IDs to monitors; used to synchronize threads during strategy execution.");
+
+    @Override
+    public List<DockerHostLocation> filterLocations(List<DockerHostLocation> locations, Entity entity) {
+        if (locations == null || locations.isEmpty()) {
+            return ImmutableList.of();
+        }
+        if (getDockerInfrastructure() == null) config().set(DOCKER_INFRASTRUCTURE, Iterables.getLast(locations).getDockerInfrastructure());
+        List<DockerHostLocation> available = MutableList.copyOf(locations);
+        boolean requireExclusive = config().get(REQUIRE_EXCLUSIVE);
+
+        try {
+            acquireMutex(entity.getApplicationId(), "Filtering locations for " + entity);
+        } catch (InterruptedException ie) {
+            Exceptions.propagate(ie); // Should never happen...
+        }
+
+        // Find hosts with entities from our application deployed there
+        Iterable<DockerHostLocation> sameApplication = Iterables.filter(available, hasApplicationId(entity.getApplicationId()));
+
+        // Check if hosts have any deployed entities that share a parent with the input entity
+        Optional<DockerHostLocation> sameParent = Iterables.tryFind(sameApplication, childrenOf(entity.getParent()));
+        if (sameParent.isPresent()) {
+            LOG.debug("Returning {} (same parent) for {} placement", sameParent.get(), entity);
+            return ImmutableList.copyOf(sameParent.asSet());
+        }
+
+        // Remove hosts if they do not have any entities from our application deployed there
+        Iterables.removeIf(available, hasApplicationId(entity.getApplicationId()));
```

Comment:
This line of code seems to do the opposite to the comment: will it not remove hosts if they *do* have any entities from our application deployed there? I presume that's what you are intending, so that things with a different parent in the same app are mutually exclusive.